R/import_metadata_and_counts.R

```diff
@@ -0,0 +1,118 @@
+#' @importFrom glue glue
+#' @importFrom rlang .data
+#' @examples
+#' data(sample_sce_obj)
+#' import_metadata_counts(sample_sce_obj,
+#'                        cache_dir = get_default_cache_dir())
+import_metadata_counts <- function(
+  sce_obj,  
+  cache_dir = get_default_cache_dir()) {
+  original_dir <- file.path(cache_dir, "original")
+  
+  # Identify metadata and counts matrix
+  metadata_tbl <- sce_obj@metadata$data
+  counts_matrix <- sce_obj@assays@data$X
+  
+  # Convert to tibble if metadata_tbl is not a tibble
+  metadata_tbl <- metadata_tbl |> as_tibble()
+  
+  # Create file_id_db from dataset_id
+  metadata_tbl <-
+    metadata_tbl |> mutate(file_id_db = .data$dataset_id |> openssl::md5() |> as.character())
+  sce_obj@metadata$data <- metadata_tbl
+  
+  # Remove existing reducedDim slot to enable get_SCE API functionality 
+  if (length(names(SingleCellExperiment::reducedDims(sce_obj))) >0 ) {
+    SingleCellExperiment::reducedDims(sce_obj) <- NULL
+  }
+  
+  # create original and cpm folders in cache directory if not exist (in order to append new counts to existing ones)
+  if (!dir.exists(original_dir)) {
+    dir.create(cache_dir, "original", recursive = TRUE)
+  }
+  
+  if (!dir.exists(file.path(cache_dir, "cpm"))) {
+    dir.create(cache_dir, "cpm", recursive = TRUE)
+  }
+  
+  # existing metadata genes
+  genes <- get_metadata() |> head(1) |> 
+    get_single_cell_experiment(cache_directory = cache_dir) |> rownames()
+  
+  # check count H5 directory name not included in the cache directory original
+  all(!metadata_tbl$file_id_db %in% dir(original_dir)) |>
+    check_true() |>
+    assert_that(msg = "Count H5 directory name should not duplicate with that in cache directory")
+
+  counts_data <- sce_obj@assays@data$X
+  assert_that(inherits(sce_obj, "SingleCellExperiment"),
+              msg = "SingleCellExperiment Object is not created from metadata.")
+  assert_that(!"^ENS" %in% genes,
+              msg = "Gene names cannot contain Ensembl IDs.")
+  assert_that(all(counts_data >= 0),
+              msg = "Counts for SingleCellExperiment cannot be negative.")
+  
+  # check the metadata contains cell_, file_id_db, sample_ with correct types
+  check_true("cell_" %in% names(metadata_tbl))
+  check_true("file_id_db" %in% names(metadata_tbl)) 
+  select(metadata_tbl, .data$cell_) |> class() |> check_character()
+  select(metadata_tbl, .data$file_id_db) |> class() |> check_character()
+  #metadata_tbl |> select(cell_, file_id_db) |> sapply(class) |> check_character()
+  
+  # check cell_ values in metadata_tbl is unique
+  (anyDuplicated(metadata_tbl$cell_) == 0 ) |> assert_that(msg = "cell_ in the metadata must be unique")
+  
+  # check cell_ values are not duplicated when join with parquet
+  cells <- select(get_metadata(), .data$cell_) |> as_tibble()
+  (!any(metadata_tbl$cell_ %in% cells$cell_)) |> assert_that(msg = "cell_ in the metadata should not duplicate with that exists in API")
```

Comment:
From the user's perspective, they aren't thinking in terms of a metadata file here, but rather in terms of the SCE object they passed in:
```suggestion
  (!any(metadata_tbl$cell_ %in% cells$cell_)) |> assert_that(msg = "Cell names (column names) should not clash with cells that already exist in the atlas")
```